File: qualification/tennison.cpp
#include <algorithm>
#include <cstdio>
#include <cstring>
#include <iostream>

#define MAXK 100
#define MAXP 1000

using namespace std;

double dp[MAXK + 1][MAXK + 1][MAXP + 1];

inline double wif(int psun, double vthen, double velse) {
  return (psun * vthen + (MAXP - psun) * velse) / MAXP;
}

int main() {
  int t; cin >> t;
  for(int tc = 1; tc <= t; tc++) {
    int k; double ps, pr, pi, pu, pw, pd, pl;
    cin >> k >> ps >> pr >> pi >> pu >> pw >> pd >> pl;
    int ipu = pu * MAXP, ipd = pd * MAXP;

    memset(dp, 0, sizeof(dp));
    dp[0][0][int(pi * MAXP)] = 1.0;

    for(int i = 0; i < k; i++) {
      for(int j = 0; j < k; j++) {
        for(int psun = 0; psun <= MAXP; psun++) {
          if(!dp[i][j][psun]) continue;
          double pwin = dp[i][j][psun] * wif(psun, ps, pr);
          if(pwin) {
            dp[i + 1][j][min(MAXP, psun + ipu)] += pwin * pw;
            dp[i + 1][j][psun] += pwin * (1.0 - pw);
          }

          double plose = dp[i][j][psun] - pwin;
          if(plose) {
            dp[i][j + 1][max(0, psun - ipd)] += plose * pl;
            dp[i][j + 1][psun] += plose * (1.0 - pl);
          }
        }
      }
    }
    double total = 0.0;
    for(int i = 0; i < k; i++) {
      for(int psun = 0; psun <= MAXP; psun++) {
        total += dp[k][i][psun];
      }
    }
    printf("Case #%d: %.6f\n", tc, total);
  }
  return 0;
}

File: qualification/tennison.scala
import java.io.{PrintStream, File}
import java.util.{Locale, Scanner}
import scala.math._

object Tennison extends App {

  Locale.setDefault(Locale.US)
  val sc = new Scanner(System.in)

  val MAXP = 1000

  @inline def wif(p: Int, vthen: BigDecimal, velse: BigDecimal) =
    (p * vthen + (MAXP - p) * velse) / MAXP

  val t = sc.nextInt()
  (1 to t).foreach { tc =>
    val k = sc.nextInt()
    val ps, pr, pi, pu, pw, pd, pl: BigDecimal = sc.nextBigDecimal()
    val ipu = (pu * MAXP).toInt
    val ipd = (pd * MAXP).toInt

    val pres = Array.fill[BigDecimal](k + 1, k + 1, MAXP + 1)(0)
    pres(0)(0)((pi * MAXP).toInt) = 1

    for(i <- 0 until k; j <- 0 until k; psun <- 0 to MAXP if pres(i)(j)(psun) > 0) {

      val pwin = pres(i)(j)(psun) * wif(psun, ps, pr)
      if(pwin > 0) {
        pres(i + 1)(j)(min(psun + ipu, MAXP)) += pwin * pw
        pres(i + 1)(j)(psun) += pwin * (1 - pw)
      }

      val plose = pres(i)(j)(psun) - pwin
      if(plose > 0) {
       pres(i)(j + 1)(max(psun - ipd, 0)) += plose * pl
       pres(i)(j + 1)(psun) += plose * (1 - pl)
      }
    }
    val total = (0 until k).map(pres(k)).map(_.sum).sum
    println("Case #%d: %.6f".format(tc, total))
  }
}
